### PR TITLE
cherry-picking db recovery #456 to master branch

### PR DESCRIPTION
**Pull request by AlexandraRoatis**
unit tests in `BlockchainDataRecoveryTest.java`

### CODE REVIEW

File: modAionImpl/src/org/aion/zero/impl/AionBlockchainImpl.java [hunk 1]
```diff
@@ -1304,60 +1311,142 @@ public IEventMgr getEventMgr() {
...
         repo.flush();
 
         // return a flag indicating if the recovery worked
-        if (repo.isValidRoot(repo.getBlockStore().getChainBlockByNumber(blockNumber).getStateRoot())) {
+        if (repo.isValidRoot(block.getStateRoot())) {
             return true;
         } else {
             // reverting back one block
-            LOG.info("Rebuild FAILED. Reverting to previous block.");
+            LOG.info("Rebuild state FAILED. Reverting to previous block.");
```
[the remark this answers is not in the document]
Comment (yao-sun) [reply modAionImpl/src/org/aion/zero/impl/AionBlockchainImpl.java]: To add to this, wouldn't this potentially change the ``bestBlock`` stored in the blockStore at runtime, while the blockchain is not informed of the change (blockchain keeps track of ``bestBlock`` in memory

File: modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java [hunk 2]
```diff
@@ -690,6 +690,50 @@ private BigInteger correctTotalDifficulty(long level, BigInteger parentTotalDiff
...
+
+            // correct block info
+            BlockInfo blockInfo = getBlockInfoForHash(levelBlocks, block.getHash());
+            if (blockInfo == null) {
+                blockInfo = new BlockInfo();
+            }
+            blockInfo.setHash(block.getHash());
+            blockInfo.setCummDifficulty(block.getDifficultyBI().add(parentTotalDifficulty));
+            // assuming side chain, with warnings upon encountered issues
+            blockInfo.setMainChain(false);
```
Comment (yao-sun) [inline modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java]: Question, whats the rationale for setting the ``mainChain`` property of the block to false? In the context of ``tryToConnectInternal`` what if a mainChain block were the one that had a corrupted index?

Comment (AlexandraRoatis) [reply modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java]: correcting the main chain flag in 3229eea

File: modAionImpl/src/org/aion/zero/impl/AionBlockchainImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modAionImpl/src/org/aion/zero/impl/AionBlockchainImpl.java]: separated the revert functionality from state recovery in d4a54fb

**Pull request merged**